### PR TITLE
Support semi concrete eval

### PR DESCRIPTION
**Pull request by ianatol**
Rebase of #280

### CODE REVIEW

**Comment by Keno** [comment]
@aviatesk can you review and merge, since this is now on master?

**Comment by aviatesk** [comment]
I'm working on this.

A general request: please make sure to add test cases when submitting a PR with a certain amount of changes/new features like this. Currently this PR yields an obvious `UndefVarError` at the `lookup` method and this sort of error could easily fixed during making up a test case.

File: src/Cthulhu.jl [hunk 1]
```diff
@@ -434,10 +435,17 @@ function _descend(term::AbstractTerminal, interp::AbstractInterpreter, curs::Abs
         """)
     end
     while true
-        if override !== nothing
+        if override !== nothing && !isa(override, SemiConcreteCallInfo)
             (; src, rt, infos, slottypes, codeinf, effects) = lookup_constproped(interp, curs, override, optimize)
         else
-            if optimize
+            if isa(override, SemiConcreteCallInfo)
```
Comment (aviatesk) [inline src/Cthulhu.jl:441]: We should lift this branch up, since otherwise the `lookup` call later will override the IR.

File: src/callsite.jl [hunk 2]
```diff
@@ -297,6 +306,13 @@ function show_callinfo(limiter, ci::ConstPropCallInfo)
     __show_limited(limiter, name, tt, get_rt(ignorewrappers(ci.mi)::MICallInfo))
 end
 
+function show_callinfo(limiter, ci::SemiConcreteCallInfo)
+    # XXX: The first argument could be const-overriden too
+    name = get_mi(ci).def.name
+    tt = ci.argtypes[2:end]
+    __show_limited(limiter, name, tt, get_rt(ci))
```
Comment (aviatesk) [inline src/callsite.jl:313]: Now we need to pass `effects`

**Comment by Keno** [comment]
Thanks!

**Pull request merged**